File: frontend/src/models/baseline.ts
// models/baseline.ts
// Baseline, RuleSet, RuleDefinition, RuleOverride model types

import { SeverityLevel } from './profile';

export type BaselineStatus = 'published' | 'draft' | 'deprecated';

export interface IdentificationStrategy {
  method: string;
  id_fields: string[];
}

export interface RuleOverride {
  [param_key: string]: number | string | boolean;
}

export interface RuleDefinition {
  id: string;
  name: string;
  enabled: boolean;
  condition: string;       // read-only DSL string
  threshold: string;
  category: string;
  severity: SeverityLevel;
  parameter_profile_id: string;
  threshold_profile_id: string;
  scope_selector_id: string;
  parameters: Record<string, string | number | boolean>; // profile defaults
  rule_overrides?: RuleOverride;
}

export interface RuleSet {
  ruleset_id: string;
  name: string;
  description: string;
  priority: number;
  rule_ids: string[];
}

export interface Baseline {
  id: string;
  name: string;
  version: string;
  status: BaselineStatus;
  description: string;
  created_at: string;
  published_at: string | null;
  rule_count: number;
  enabled_count: number;
  identification_strategy: IdentificationStrategy;
  naming_template: string;
  parameter_profile_id?: string;
  threshold_profile_id?: string;
  scope_selector_id?: string;
  ruleset_ids?: string[];
  // legacy flat params kept for backward compat with prototype
  parameters?: Record<string, string>;
}

File: backend/models/baseline.py
# backend/models/baseline.py
# Baseline-related models

from typing import Optional, Union, List
from pydantic import BaseModel


class ParameterProfile(BaseModel):
    profile_id: str
    name: str
    description: Optional[str] = None
    parameters: dict


class ThresholdProfile(BaseModel):
    profile_id: str
    name: str
    description: Optional[str] = None
    thresholds: dict


class ScopeSelector(BaseModel):
    scope_id: str
    name: str
    description: Optional[str] = None
    selector_type: str


class RuleSet(BaseModel):
    ruleset_id: str
    name: str
    description: Optional[str] = None
    rule_ids: List[str] = []


class ParameterDefinition(BaseModel):
    name: str
    type: str
    default: Optional[Union[float, str, bool]] = None
    unit: Optional[str] = None
    description: Optional[str] = None


class ConditionExpression(BaseModel):
    expression: str
    description: Optional[str] = None


class RuleDefinition(BaseModel):
    id: str
    name: str
    enabled: bool = True
    category: str
    severity: str
    condition: ConditionExpression
    parameters: List[ParameterDefinition] = []
    description: Optional[str] = None


class Baseline(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    version: str
    status: str = "draft"
    identification_strategy: dict = {}
    naming_template: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
